stockflux-launcher/src/free-text-search/FreeTextSearch.js

```diff
@@ -0,0 +1,216 @@
+import { createWindow, populateResultsContainer } from './SearchResultWindow';
+import './FreeTextSearch.css';
+
+const SEARCHING = 'searching';
+const SUCCESS = 'success';
+const ERROR = 'error';
+const INITIALISE = 'initialise';
+const SEARCH_TIMEOUT_INTERVAL = 250;
+const NO_MATCHES_HTML = <p>Sorry, no matches found.</p>;
+const INITIAL_MESSAGE_HTML = (
+  <p>
+    Use the input field above to search for instruments or click on the search
+    icon to close the search.
+  </p>
+);
+
+let latestRequest = null;
+let resultsWindow = null;
+
+const initialSearchState = {
+  isSearching: false,
+  hasErrors: false
+};
+
+const searchReducer = (state, { type, results }) => {
+  switch (type) {
+    case SEARCHING:
+      return {
+        ...state,
+        hasErrors: false,
+        isSearching: true
+      };
+    case SUCCESS:
+      return {
+        ...state,
+        isSearching: false,
+        results
+      };
+    case ERROR:
+      return {
+        ...state,
+        hasErrors: true,
+        isSearching: false
+      };
+    case INITIALISE:
+      return initialSearchState;
+    default:
+      throw new Error();
+  }
+};
+
+const Search = props => {
+  const [searchState, dispatch] = useReducer(searchReducer, initialSearchState);
+  const [parentUuid, setParentUuid] = useState(null);
+  const [query, setQuery] = useState(null);
+  const [windowState] = WindowHooks.useCurrentWindowState();
+  const [debouncedQuery, setDebouncedQuery] = useState(null);
+  const { isSearching, results } = searchState;
+  const searchButtonRef = useRef(null);
+  const searchInputRef = useRef(null);
+
+  const setResultsWindow = newReference => {
+    resultsWindow = newReference;
+  };
+
+  const resetLauncherInputField = () => {
+    if (searchInputRef.current) searchInputRef.current.value = '';
+  };
+
+  const closeResultsWindow = useCallback(() => {
+    if (resultsWindow) {
+      resultsWindow.close();
+      setResultsWindow(null);
+    }
+  }, []);
+
+  const handleOnInputChange = useCallback(
+    event => {
+      setQuery(event.target.value);
+      if (event.target.value !== null && event.target.value.length === 0) {
+        closeResultsWindow();
+      } else if (!resultsWindow)
+        createWindow(
+          searchButtonRef,
+          searchInputRef,
+          props.dockedTo,
+          windowState.bounds,
+          setResultsWindow
+        ).catch(err => console.error(err));
+    },
+    [closeResultsWindow, props.dockedTo, windowState.bounds]
+  );
+
+  const handleSearchClick = useCallback(async () => {
+    if (resultsWindow) {
+      closeResultsWindow();
+      if (props.dockedTo === Constants.ScreenEdge.TOP)
+        resetLauncherInputField();
+    } else {
+      createWindow(
+        searchButtonRef,
+        searchInputRef,
+        props.dockedTo,
+        windowState.bounds,
+        setResultsWindow
+      ).catch(err => console.error(err));
+    }
+  }, [closeResultsWindow, props.dockedTo, windowState.bounds]);
```

Comment:
Does this callback need to be rebound whenever searchButtonRef or searchInputRef updates?